apis/Google.Cloud.Logging.NLog/Google.Cloud.Logging.NLog/GoogleCloudLoggingTarget.cs

```diff
@@ -0,0 +1,389 @@
+        {
+            GoogleCredential credential = GetCredentialFromFile();
+            if (credential == null)
+            {
+                return LoggingServiceV2Client.Create();
+            }
+
+            Grpc.Core.Channel channel = new Grpc.Core.Channel(
+                LoggingServiceV2Client.DefaultEndpoint.Host,
+                LoggingServiceV2Client.DefaultEndpoint.Port,
+                Grpc.Auth.GoogleGrpcCredentials.ToChannelCredentials(credential)
+            );
+
+            return LoggingServiceV2Client.Create(channel);
+        }
+
+        private GoogleCredential GetCredentialFromFile()
+        {
+            var credentialFile = CredentialFile?.Render(LogEventInfo.CreateNullEvent());
+            if (string.IsNullOrWhiteSpace(credentialFile))
+            {
+                return null;
+            }
+
+            GoogleCredential credential = GoogleCredential.FromFile(credentialFile);
+
+            if (credential.IsCreateScopedRequired)
+            {
+                credential = credential.CreateScoped(s_oAuthScopes);
+            }
+
+            return credential;
+        }
+
+        private void ActivateLogIdAndResource(string logId)
+        {
+            string projectId = null;
+            MonitoredResource resource = null;
+            if (!DisableResourceTypeDetection)
+            {
+                resource = MonitoredResourceBuilder.FromPlatform(_platform);
+                resource.Labels.TryGetValue("project_id", out projectId);
+            }
+
+            if (projectId == null)
+            {
+                // Either platform detection is disabled, or it detected an unknown platform
+                // So use the manually configured projectId and override the resource
+                projectId = GaxPreconditions.CheckNotNull(ProjectId?.Render(LogEventInfo.CreateNullEvent()), nameof(ProjectId));
+                if (ResourceType == null)
+                {
+                    resource = new MonitoredResource { Type = "global", Labels = { { "project_id", projectId } } };
+                }
+                else
+                {
+                    resource = new MonitoredResource { Type = ResourceType,
+                        Labels = { ResourceLabels.ToDictionary(x => x.Name, x => x.Layout.Render(LogEventInfo.CreateNullEvent())) } };
+                }
+            }
+
+            _resource = resource;
+            var logName = new LogName(projectId, logId);
+            _logName = logName.ToString();
+            _logNameToWrite = LogNameOneof.From(logName);
+        }
+
+        /// <summary>
+        /// 
+        /// </summary>
+        protected override void Write(AsyncLogEventInfo logEvent)
+        {
+            var continuation = logEvent.Continuation;
+            var logEntry = BuildLogEntry(logEvent.LogEvent);
+            _prevTask = _client.WriteLogEntriesAsync(_logNameToWrite, _resource, s_emptyLabels, new[] { logEntry }, _cancelTokenSource.Token)

```

Comment:
@snakefoot What happens if this write fails? Are the log entries permanently lost, or does NLog contain retry logic? If not, what would happen if we put a retry with backoff here?

Reply:
One can use https://github.com/NLog/NLog/wiki/RetryingWrapper-target to enable retry. There is also https://github.com/nlog/NLog/wiki/FallbackGroup-target to use alternative target if this fails.

Reply:
Yes the retry-wrapper just need to be placed before the async-wrapper. So it becomes retry-wrapper -> async-wrapper -> google-stack-driver-target.